usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_sis.c

```diff
@@ -0,0 +1,251 @@
+		if ((doorbell & SIS_REENABLE_SIS_MODE) == 0) {
+			return (B_TRUE);
+		}
+		drv_usecwait(1000);	/* ---- Wait 1ms ---- */
+	}
+	return (B_FALSE);
+}
+
+boolean_t
+sis_wait_for_ctrl_ready(pqi_state_t s)
+{
+	int		loop_count;
+	uint32_t	status;
+
+	for (loop_count = 0; loop_count < 1000; loop_count++) {
+		status = G32(s, sis_firmware_status);
+		if (status & SIS_CTRL_KERNEL_PANIC)
+			return (B_FALSE);
+		if (status & SIS_CTRL_KERNEL_UP)
+			return (B_TRUE);
+		drv_usecwait(1000);	 /* ---- Wait 1ms ---- */
+	}
+	return (B_FALSE);
+}
+
+/*
+ * sis_get_ctrl_props -- Verify we're talking to controller that speaks PQI
+ */
+boolean_t
+sis_get_ctrl_props(pqi_state_t s)
+{
+	sis_sync_cmd_params_t	p;
+	uint32_t		property;
+	uint32_t		extended_property;
+
+	(void) memset(&p, 0, sizeof (p));
+	if (sis_send_sync_cmd(s, SIS_CMD_GET_ADAPTER_PROPERTIES, &p) == B_FALSE)
+		return (B_FALSE);
+
+	property = p.mailbox[1];
+	if (!(property & SIS_EXTENDED_PROPERTIES_SUPPORTED))
+		return (B_FALSE);
+
+	extended_property = p.mailbox[4];
+	if ((extended_property & SIS_REQUIRED_EXTENDED_PROPERTIES) !=
+	    SIS_REQUIRED_EXTENDED_PROPERTIES)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+boolean_t
+sis_get_pqi_capabilities(pqi_state_t s)
+{
+	sis_sync_cmd_params_t	p;
+
+	(void) memset(&p, 0, sizeof (p));
+	if (sis_send_sync_cmd(s, SIS_CMD_GET_PQI_CAPABILITIES, &p) == B_FALSE)
+		return (B_FALSE);
+
+	s->s_max_sg_entries = p.mailbox[1];
+	s->s_max_xfer_size = p.mailbox[2];
+	s->s_max_outstanding_requests = p.mailbox[3];
+	s->s_config_table_offset = p.mailbox[4];
+	s->s_config_table_len = p.mailbox[5];
+	return (B_TRUE);
+}
+
+#define	SIS_ROUNDUP(a, align, type) \
+	(type)((((uintptr_t)(a) + align - 1) & ~align))
+
+boolean_t
+sis_init_base_struct_addr(pqi_state_t s)
+{
+	sis_base_struct_t	*base;
+	pqi_dma_overhead_t	*o;
+	sis_sync_cmd_params_t	params;
+	boolean_t		rc;
+	void			*dma_addr;
+
+	o = pqi_alloc_single(s, sizeof (*base) + SIS_BASE_STRUCT_ALIGNMENT);
+	if (o == NULL)
+		return (B_FALSE);
+
+	base = PQIALIGN_TYPED(o->alloc_memory, SIS_BASE_STRUCT_ALIGNMENT,
+	    sis_base_struct_t *);
+	base->sb_revision = SIS_BASE_STRUCT_REVISION;
+	base->sb_error_buffer_paddr_low = (uint32_t)s->s_error_dma->dma_addr;
+	base->sb_error_buffer_paddr_high =
+	    (uint32_t)(s->s_error_dma->dma_addr >> 32);
+	base->sb_error_elements_len = PQI_ERROR_BUFFER_ELEMENT_LENGTH;
+	base->sb_error_elements_num = s->s_max_outstanding_requests;
+
+	dma_addr = PQIALIGN_TYPED(o->dma_addr, SIS_BASE_STRUCT_ALIGNMENT,
+	    void *);
+	(void) memset(&params, 0, sizeof (params));
+	params.mailbox[1] = (uint32_t)(intptr_t)dma_addr;
+	params.mailbox[2] = (uint32_t)((uint64_t)((intptr_t)dma_addr) >> 32);
+	params.mailbox[3] = sizeof (*base);
+	(void) ddi_dma_sync(o->handle, 0, 0, DDI_DMA_SYNC_FORDEV);
+	rc = sis_send_sync_cmd(s, SIS_CMD_INIT_BASE_STRUCT_ADDRESS, &params);
+
+	pqi_free_single(s, o);
+
+	return (rc);
+}
+
+/*
+ * []------------------------------------------------------------[]
```

Comment:
Spelling in the comment below, and just as a note, I personally find these cutesy style block comments distracting and annoying -- as soon as you adjust the text the alignment gets all screwed up, and I'd rather not waste time counting hyphens.